# Final Project: Crypto Currency from 2013 to 2021

Member of our team:

<table style='float:center'>
    <tr>
        <th>Fullname</th>
        <th>Student ID</th>
    </tr>
    <tr>
        <td>Đoàn Ánh Dương</td>
        <td>20127474</td>
    </tr>
</table>




---

## Import Library

Necessary file imported here.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

import os
import glob

Environment using within this final project.

In [3]:
import sys
sys.executable

'/home/duong/anaconda3/envs/min_ds-env/bin/python'

Set options here.

In [2]:
pd.set_option('display.max_colwidth', None)
pd.set_option('display.float_format', '{:.2f}'.format)

---

## I. Idea and Expectation
### Idea
For recent years, we have been heard at least once about crypto, particuarly Bitcoin and its application. Although digital currency brought many controversial opinions up to now, we cannot realize its success and popularity. 

So I decided to deep in crypto market and fortunately, I found the historical price data of crypto currency on Kaggle. Although the data does not contain all information about a specific coin, the historical price can bring us to take an overview of digital currency evolution.

Throughout this project, I hope that we can have a perspective objectively and a basic knowledge about digital currency.
### Expectation
In this project, we will find out the crypto currency market and its history, with expectation that we can find a method to buy a coin and take profit. 

## II. Datasets and Collection

### Source and Public Data
I found the datasets used in this project was public on [Kaggle](https://www.kaggle.com/datasets/sudalairajkumar/cryptocurrencypricehistory): Crypto Currrency price history.

### License
The data is taken from [coinmarketcap](https://coinmarketcap.com/) and it is [free](https://coinmarketcap.com/faq/) to use the data.

More about the license I found on kaggle: 
- CC0: Public Domain ([No copyright](https://creativecommons.org/publicdomain/zero/1.0/))

### About Data
The dataset has one csv file for each currency. **Price history is available on a daily basis from <font color = 'green'>*April 28, 2013*</font>**. This dataset has the historical price information of some of the top crypto currencies by market capitalization, including Bitcoin, Etherium, Binance Coin,...

### Description 
All information of columns contain in the file named `description.txt`.

### Author
Datasets were collected by Sudalai Rajkumar(registered account name [SRK](https://www.kaggle.com/sudalairajkumar)) on Kaggle.

### How did the author collect data?
As the author's answer, there is not mentioned clearly how the way he collected data, maybe collected by scraping historical data web. For instance: [bitcoin](https://coinmarketcap.com/currencies/bitcoin/historical-data/).

### Is dataset subjective or objective?
Besides the historical price of digital currency, there are many factors in a coin we need to take consideration on. For example: Holders Statistic over time, legality, team and organization investors, who created the coin... 

So that the dataset is subjective.

---

## III. Preprocessing Data

### 1. Change files.

First, we cannot access each files to analyze, it must be combined into one.

#### How many files does we have?

In [4]:
path = 'archive/'
os.chdir(path)
ext = 'csv'
all_files = [f for f in glob.glob(f'*.{ext}')]

print(f'Containing number of files: {len(all_files)}.')
print(all_files)

os.chdir('../')

Containing number of files: 23.
['coin_XRP.csv', 'coin_Ethereum.csv', 'coin_Aave.csv', 'coin_USDCoin.csv', 'coin_Solana.csv', 'coin_EOS.csv', 'coin_WrappedBitcoin.csv', 'coin_Cardano.csv', 'coin_NEM.csv', 'coin_Tron.csv', 'coin_CryptocomCoin.csv', 'coin_BinanceCoin.csv', 'coin_Iota.csv', 'coin_ChainLink.csv', 'coin_Monero.csv', 'coin_Dogecoin.csv', 'coin_Tether.csv', 'coin_Bitcoin.csv', 'coin_Cosmos.csv', 'coin_Litecoin.csv', 'coin_Stellar.csv', 'coin_Polkadot.csv', 'coin_Uniswap.csv']


#### Are there duplicated in each files? 

Lets check duplicated rows.

In [5]:
for f in all_files:
    temp = pd.read_csv(path+f)
    print(f'duplicated in {f}:', temp.index.duplicated().sum())

duplicated in coin_XRP.csv: 0
duplicated in coin_Ethereum.csv: 0
duplicated in coin_Aave.csv: 0
duplicated in coin_USDCoin.csv: 0
duplicated in coin_Solana.csv: 0
duplicated in coin_EOS.csv: 0
duplicated in coin_WrappedBitcoin.csv: 0
duplicated in coin_Cardano.csv: 0
duplicated in coin_NEM.csv: 0
duplicated in coin_Tron.csv: 0
duplicated in coin_CryptocomCoin.csv: 0
duplicated in coin_BinanceCoin.csv: 0
duplicated in coin_Iota.csv: 0
duplicated in coin_ChainLink.csv: 0
duplicated in coin_Monero.csv: 0
duplicated in coin_Dogecoin.csv: 0
duplicated in coin_Tether.csv: 0
duplicated in coin_Bitcoin.csv: 0
duplicated in coin_Cosmos.csv: 0
duplicated in coin_Litecoin.csv: 0
duplicated in coin_Stellar.csv: 0
duplicated in coin_Polkadot.csv: 0
duplicated in coin_Uniswap.csv: 0


So there is no duplicated rows. We carry out combining files.

In [6]:
combined_csv = pd.concat([pd.read_csv(path + fname) for fname in all_files])
print('duplicated in combine files: ', combined_csv.index.duplicated().sum())

duplicated in combine files:  34091


Why the combined files take duplicated? Because column 'SNo' is the index starting with $1$ for each files. So we must reset the index in combined files. 

In [7]:
combined_csv.drop('SNo',1,inplace = True)
combined_csv.to_csv('crypto_finance.csv', index=False)

combined_csv = pd.read_csv('crypto_finance.csv')
combined_csv.index.rename('SNo', inplace = True)
combined_csv.to_csv('crypto_finance.csv')

### 2. Pre-processing

Read new file. Let's see some values.

In [3]:
cryp_df = pd.read_csv('crypto_finance.csv', index_col='SNo')

cryp_df.sample(10)

Name Symbol                 Date   High    Low   Open  Close  \
SNo                                                                       
4196   Ethereum    ETH  2019-03-03 23:59:59 135.85 131.33 134.79 132.25   
24307  Dogecoin   DOGE  2020-03-03 23:59:59   0.00   0.00   0.00   0.00   
34098   Stellar    XLM  2015-01-08 23:59:59   0.01   0.01   0.01   0.01   
25380    Tether   USDT  2016-10-05 23:59:59   1.00   1.00   1.00   1.00   
20954    Monero    XMR  2018-07-19 23:59:59 143.68 138.55 139.60 139.53   
35357   Stellar    XLM  2018-06-20 23:59:59   0.24   0.23   0.24   0.23   
20693    Monero    XMR  2017-10-31 23:59:59  89.33  87.60  88.75  87.64   
12869      TRON    TRX  2017-11-24 23:59:59   0.00   0.00   0.00   0.00   
26595    Tether   USDT  2020-02-02 23:59:59   1.01   1.00   1.00   1.00   
1043        XRP    XRP  2016-06-13 23:59:59   0.01   0.01   0.01   0.01   

              Volume      Marketcap  
SNo                                  
4196   3454374052.60 13900679027.02  
24307   174988478.67   308838196.56  
34098      222723.00    19828782.80  
25380      753062.00     6951590.95  
20954    30172600.00  2266122952.65  
35357    43000600.00  4320819301.91  
20693    28704900.00  1340360432.37  
12869     4204470.00   132859497.29  
26595 40560625161.10  4644893208.29  
1043      1537720.00   203072051.01

More information about dataset.

In [13]:
cryp_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 37082 entries, 0 to 37081
Data columns (total 9 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Name       37082 non-null  object        
 1   Symbol     37082 non-null  object        
 2   Date       37082 non-null  datetime64[ns]
 3   High       37082 non-null  float64       
 4   Low        37082 non-null  float64       
 5   Open       37082 non-null  float64       
 6   Close      37082 non-null  float64       
 7   Volume     37082 non-null  float64       
 8   Marketcap  37082 non-null  float64       
dtypes: datetime64[ns](1), float64(6), object(2)
memory usage: 3.8+ MB


#### How many rows and columns in the file?

In [4]:
cryp_df.shape

(37082, 9)

There are 37082 rows and 9 columns.

#### What is the meaning of each row (sample)?

The dataset shows that each samples contain the information of a specific coin at each time.

#### Does the data have duplicated samples?

In [5]:
cryp_df.index.duplicated().sum()

0

There is no duplicated samples.

#### What does each column mean? 

In [6]:
f = open('description.txt', 'r')
print(f.read())
f.close()

1) SNo:         Serial Number.
2) Name:        Name of coin.
3) Symbol:      Symbol of coin.
4) Date:        Date of observation.
5) Open:        Opening price on the given day.
6) High:        Highest price on the given day.
7) Low:         Lowest price on the given day.
8) Close:       Closing price on the given day.
9) Volume:      Volume of transactions on the given day (in USD).
10) Market Cap: Market capitalization in USD.


#### Check the data types? 

In [8]:
cryp_df.dtypes

Name          object
Symbol        object
Date          object
High         float64
Low          float64
Open         float64
Close        float64
Volume       float64
Marketcap    float64
dtype: object

Overall, all the column types is correct except `Date`. We need to change `Date` column to appropriate type (`datetime`).

In [11]:
cryp_df['Date'] = pd.to_datetime(cryp_df['Date'], format="%Y-%m-%d %X")

#### With the numeric columns, how are values distributed? Are there any abnormal values?

In [31]:
nume_cols_df = cryp_df.describe(include=['number', 'datetime'], datetime_is_numeric=True)
nume_cols_df.loc['missing_ratio'] = cryp_df.agg(lambda x: x.isna().mean()*100)

nume_cols_df

Date     High      Low     Open  \
count                                  37082 37082.00 37082.00 37082.00   
mean           2018-08-16 07:12:30.221616896  1016.06   952.99   985.32   
min                      2013-04-29 23:59:59     0.00     0.00     0.00   
25%                      2017-03-05 23:59:59     0.08     0.07     0.07   
50%                      2019-01-09 23:59:59     1.01     1.00     1.00   
75%                      2020-05-13 23:59:59    31.92    29.00    30.46   
max                      2021-07-06 23:59:59 64863.10 62208.96 63523.75   
std                                      NaN  5249.50  4907.93  5088.10   
missing_ratio                           0.00     0.00     0.00     0.00   

                 Close          Volume        Marketcap  
count         37082.00        37082.00         37082.00  
mean            987.12   3022541604.27   15429426181.11  
min               0.00            0.00             0.00  
25%               0.07      4937190.00     239595503.59  
50%               1.00     85128048.00    1405334561.55  
75%              30.51    938848893.11    5159304767.10  
max           63503.46 350967941479.06 1186364044140.27  
std            5093.70  11909631165.50   70591279351.79  
missing_ratio     0.00            0.00             0.00

Fortunately, there is no NaN values in DataFrame. But it exists abnormal values in `min` row.

So, lets take a look at these columns.